cf-agent/package_module.c

```diff
@@ -58,21 +62,53 @@ PackageModuleWrapper *NewPackageModuleWrapper(PackageModuleBody *package_module)
 
     PackageModuleWrapper *wrapper = xmalloc(sizeof(PackageModuleWrapper));
 
-    wrapper->path = GetPackageModuleRealPath(package_module->name);
+    GetPackageModuleExecInfo(package_module, &(wrapper->path),
+                             &(wrapper->script_path), &(wrapper->script_exec_opts));
     wrapper->name = SafeStringDuplicate(package_module->name);
     wrapper->package_module = package_module;
 
-    /* Check if file exists */
-    if (!wrapper->path || (access(wrapper->path, X_OK) != 0))
+    if (wrapper->path == NULL)
     {
-        Log(LOG_LEVEL_ERR,
-            "can not find package wrapper in provided location '%s' "
-            "or access to file is restricted: %d",
-            wrapper->path, errno);
+        Log(LOG_LEVEL_ERR, "No executable for the package module '%s'", package_module->name);
         DeletePackageModuleWrapper(wrapper);
         return NULL;
     }
 
+    /* Check if file(s) exist(s) */
+    if (wrapper->script_path != NULL)
+    {
+        if (access(wrapper->path, X_OK) != 0)
+        {
+            Log(LOG_LEVEL_ERR,
+                "can not find package wrapper interpreter in provided location '%s' "
+                "or access to file is restricted: %s",
+                wrapper->path, strerror(errno));
+            DeletePackageModuleWrapper(wrapper);
+            return NULL;
+        }
+        if (access(wrapper->script_path, R_OK) != 0)
+        {
+            Log(LOG_LEVEL_ERR,
+                "can not find the package wrapper script in provided location '%s' "
+                "or access to file is restricted: %s",
+                wrapper->script_path, strerror(errno));
+            DeletePackageModuleWrapper(wrapper);
+            return NULL;
+        }
+    }
+    else
+    {
+        if (access(wrapper->path, X_OK) != 0)
+        {
+            Log(LOG_LEVEL_ERR,
+                "can not find package wrapper in provided location '%s' "
+                "or access to file is restricted: %s",
+                wrapper->path, strerror(errno));
+            DeletePackageModuleWrapper(wrapper);
+            return NULL;
+        }
```

Comment:
Add a comment or log message here; What does it mean when `script_path == NULL` and `access() == 0`?

Reply:
There are log messages. What do you mean? The code here is just checking that files exist and have proper permissions set. If something is wrong, an error is logged and `NULL` is returned otherwise a proper `PackageModuleWrapper` is returned.

Reply:
I just think these ifs are a little hard to decode (especially because of `wrapper->script_path` and `wrapper->path`). I'd add a comment along the lines of:

```
// Access for both script and interpreter is OK
```

Making variables for these two, might also improve readability:

```
const char *script_path = wrapper->script_path;
const char *interpreter_path = wrapper->path;
```

Reply:
Done.